Abstractions and shortestpathrouter

dakk opened the pull request:
- abstracting iso creation function for different criteria (not only wind routing)
- shortestpathrouter for fixedSpeed routing (aka motoring)
- nautical mile precision (a precise constant for conversion, avoiding magic numbers in code)
- avoid speed calculation if present in isochrone (there was a redundant calculation for speed)

@enricofer please check this pull, it does not break tests (expect for values depending on nautical mile precision)

enricofer commented on tests/linearbestisorouter_test.py:
```diff
@@ -101,11 +55,10 @@ def test_step(self):
 
         path_to_end = res.path + [[*self.track[-1],'']]
         self.assertEqual( res.time, datetime.datetime.fromisoformat('2021-04-02 19:00:00'))
-        self.assertEqual(len(json.dumps(weatherrouting.utils.pathAsGeojson(path_to_end))), 2839)
```
this is a very rough way to check if the returned path is the same.
what about checking  geojson identity with md5 hash?

```
import hashlib
md5_res = hashlib.md5(json.dumps(weatherrouting.utils.pathAsGeojson(path_to_end)).encode())
self.assertEqual(md5_res.hexdigest(), 'f5f3ec1f021679c3009865bc162782e0')
```

dakk replied:
yep, good idea; I'm using sha2 instead of md5

enricofer commented on weatherrouting/polar.py:
```diff
@@ -171,8 +171,9 @@ def getTWARoutage(self,tws,twa):
 
 		
-	def maxReachDistance(self,p,twd,tws):
+	def maxReachDistance(self, p, twd, tws, speed = None):
```
It seems this function can be simplified in 
```
	def maxReachDistance(self, p, speed):
		dt = (1. / 60. * 60.)
		maxp = utils.routagePointDistance (p[0], p[1], speed * dt, 1)
		return utils.pointDistance(p[0], p[1], maxp[0], maxp[1])
```

dakk replied:
Ok

enricofer commented on weatherrouting/routers/linearbestisorouter.py:
```diff
@@ -51,7 +50,7 @@ def generate_path(p):
 				if distance_to_end_point < self.getParamValue('minIncrease'):
 					(twd,tws) = self.grib.getWindAt (time + datetime.timedelta(hours=1), p[0], p[1])
-					maxReachDistance = self.polar.maxReachDistance(p, twd, tws)
+					maxReachDistance = self.polar.maxReachDistance(p, twd, tws, p[6])
```
see above
```
maxReachDistance = self.polar.maxReachDistance(p, p[6])
```

enricofer reviewed (commented):
For me is ok, anyway I suggest a code clean-up and tests enhancement

dakk commented:
> For me is ok, anyway I suggest a code clean-up and tests enhancement

Do you have any suggestions about it? 


In the last commit I made some modifications:
- since maxReachDistance does not use polar anymore, I moved it to utils
- updated all maxReachDistance calls and moved test to utils
- generalized route function for isochrone routing; now shortestpathrouter extends linearbestisorouter, so no common code is duplicated

enricofer commented:
Ok. It seems clean.
